tenant_schemas/middleware.py

```diff
@@ -13,38 +13,39 @@
 else:
     MIDDLEWARE_MIXIN = object
 
+"""
+These middlewares should be placed at the very top of the middleware stack.
+Selects the proper database schema using request information. Can fail in
+various ways which is better than corrupting or revealing data.
+
+Extend BaseTenantMiddleware for a custom tenant selection strategy,
+such as inspecting the header, or extracting it from some OAuth token.
+"""
+
+class BaseTenantMiddleware(MIDDLEWARE_MIXIN):
 
-class TenantMiddleware(MIDDLEWARE_MIXIN):
     """
-    This middleware should be placed at the very top of the middleware stack.
-    Selects the proper database schema using the request host. Can fail in
-    various ways which is better than corrupting or revealing data.
+    Subclass and override  this to achieve desired behaviour. Given a
+    request, return the tenant to use. Tenant should be an instance
+    of TENANT_MODEL.
     """
-    TENANT_NOT_FOUND_EXCEPTION = Http404
-
-    def hostname_from_request(self, request):
-        """ Extracts hostname from request. Used for custom requests filtering.
-            By default removes the request's port and common prefixes.
-        """
-        return remove_www(request.get_host().split(':')[0]).lower()
+    def get_tenant(self, request):
+        raise Exception("You have not implemented get_tenant. Don't know which tenant to use!")
```

Comment:
I agree. Done in https://github.com/bernardopires/django-tenant-schemas/pull/452/commits/540f32cab6bf756a76887461a8432d49f09ef217